ICE: `downgrade_to_delayed_bug: cannot downgrade Warning(None) to DelayedBug: not an error`

<!--
[31mICE[0m: Rustc ./9AD9254BE55C2DE4C10AF9E2DCDEF6481734FCE35AE4E4E4365FC76935FC986C.rs '' 'thread 'rustc' panicked at compiler/rustc_hir_analysis/src/astconv/lint.rs:127:22: 'downgrade_to_delayed_bug: cannot downgrade Warning(None) to DelayedBug: not an error'', 'thread 'rustc' panicked at compiler/rustc_hir_analysis/src/astconv/lint.rs:127:22: 'downgrade_to_delayed_bug: cannot downgrade Warning(None) to DelayedBug: not an error''
File: /tmp/im/9AD9254BE55C2DE4C10AF9E2DCDEF6481734FCE35AE4E4E4365FC76935FC986C.rs
-->

auto-reduced (treereduce-rust):
````rust
fn call_this<F>(f: F) : Fn(&str) + call_that {}
````

original:
````rust
// Any copyright is dedicated to the Public Domain.
// http://creativecommons.org/publicdomain/zero/1.0/

// pp-exact

fn call_it(f: Box<FnMut(f: Box<FnMut(String) -> String>) -> String>) {}

fn call_this<F>(f: F) : Fn(&str) + call_that {}

fn call_that<F>(f: F) where F: for<'a> Fn(&'a isize, &'a isize) -> isize {}

fn call_extern(f: fn() -> isize) {}

fn call_abid_extern(call_it: extern "C" fn() -> isize) {}

pub fn main(&str) {}

````

Version information
````
rustc 1.77.0-nightly (11035f9f5 2024-01-05)
binary: rustc
commit-hash: 11035f9f52c65739d9613f5b9ecd332e9e640e71
commit-date: 2024-01-05
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error: return types are denoted using `->`
 --> /tmp/icemaker_global_tempdir.ZcuHjInOmhfr/rustc_testrunner_tmpdir_reporting.BaGXrEBoFcyz/mvce.rs:8:23
  |
8 | fn call_this<F>(f: F) : Fn(&str) + call_that {}
  |                       ^ help: use `->` instead

error[E0405]: cannot find trait `call_that` in this scope
 --> /tmp/icemaker_global_tempdir.ZcuHjInOmhfr/rustc_testrunner_tmpdir_reporting.BaGXrEBoFcyz/mvce.rs:8:36
  |
8 | fn call_this<F>(f: F) : Fn(&str) + call_that {}
  |                                    ^^^^^^^^^ not found in this scope

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.ZcuHjInOmhfr/rustc_testrunner_tmpdir_reporting.BaGXrEBoFcyz/mvce.rs:8:48
  |
8 | fn call_this<F>(f: F) : Fn(&str) + call_that {}
  |                                                ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.ZcuHjInOmhfr/rustc_testrunner_tmpdir_reporting.BaGXrEBoFcyz/mvce.rs`

thread 'rustc' panicked at compiler/rustc_hir_analysis/src/astconv/lint.rs:127:22:
downgrade_to_delayed_bug: cannot downgrade Warning(None) to DelayedBug: not an error
stack backtrace:
   0:     0x7f302d38b806 - std::backtrace_rs::backtrace::libunwind::trace::h72ed163bf54ed335
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7f302d38b806 - std::backtrace_rs::backtrace::trace_unsynchronized::ha9eeffcac0159741
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f302d38b806 - std::sys_common::backtrace::_print_fmt::hd878d288400665ec
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7f302d38b806 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hee1d8c1810fab8e4
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f302d3ddf00 - core::fmt::rt::Argument::fmt::he899d11a584078ee
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/core/src/fmt/rt.rs:142:9
   5:     0x7f302d3ddf00 - core::fmt::write::h9fb58a907feb62f5
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/core/src/fmt/mod.rs:1120:17
   6:     0x7f302d37f15f - std::io::Write::write_fmt::h7e0c3492389aed93
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/io/mod.rs:1810:15
   7:     0x7f302d38b5e4 - std::sys_common::backtrace::_print::hb4e839736d952adb
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f302d38b5e4 - std::sys_common::backtrace::print::ha846f6638d2e1501
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f302d38e377 - std::panicking::default_hook::{{closure}}::hcd8a486cb84a6ace
  10:     0x7f302d38e0d9 - std::panicking::default_hook::hfa2ffdc3f2a8ce4a
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/panicking.rs:292:9
  11:     0x7f3030160cac - std[735fb5c0265dddc8]::panicking::update_hook::<alloc[2bd63a676ca9eb5]::boxed::Box<rustc_driver_impl[d85864648598caad]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f302d38eac6 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hc9ef68fbc75324a4
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/alloc/src/boxed.rs:2030:9
  13:     0x7f302d38eac6 - std::panicking::rust_panic_with_hook::h869d110b23694b5d
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/panicking.rs:783:13
  14:     0x7f302d38e812 - std::panicking::begin_panic_handler::{{closure}}::hd3a60d30fcbcee10
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/panicking.rs:657:13
  15:     0x7f302d38bd06 - std::sys_common::backtrace::__rust_end_short_backtrace::hcea5d524563d6535
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7f302d38e570 - rust_begin_unwind
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/panicking.rs:645:5
  17:     0x7f302d3da605 - core::panicking::panic_fmt::hc641175b49629a2b
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/core/src/panicking.rs:72:14
  18:     0x7f302f7c4858 - <rustc_errors[5a03629f0326f870]::diagnostic::Diagnostic>::downgrade_to_delayed_bug
  19:     0x7f303028382f - <dyn rustc_hir_analysis[4501648514be1474]::astconv::AstConv>::maybe_lint_impl_trait
  20:     0x7f303023d6ec - <<dyn rustc_hir_analysis[4501648514be1474]::astconv::AstConv>::maybe_lint_bare_trait::{closure#1} as core[d0d16022f048c5c1]::ops::function::FnOnce<(&mut rustc_errors[5a03629f0326f870]::diagnostic_builder::DiagnosticBuilder<()>,)>>::call_once::{shim:vtable#0}
  21:     0x7f3030258722 - <alloc[2bd63a676ca9eb5]::boxed::Box<dyn for<'a, 'b> core[d0d16022f048c5c1]::ops::function::FnOnce<(&'a mut rustc_errors[5a03629f0326f870]::diagnostic_builder::DiagnosticBuilder<'b, ()>,), Output = ()>> as core[d0d16022f048c5c1]::ops::function::FnOnce<(&mut rustc_errors[5a03629f0326f870]::diagnostic_builder::DiagnosticBuilder<()>,)>>::call_once
  22:     0x7f303221a7e4 - rustc_middle[8ef003f1e476f553]::lint::struct_lint_level::struct_lint_level_impl::<&str>
  23:     0x7f3031601b8c - <dyn rustc_hir_analysis[4501648514be1474]::astconv::AstConv>::ast_ty_to_ty_inner
  24:     0x7f30315fed76 - <dyn rustc_hir_analysis[4501648514be1474]::astconv::AstConv>::ty_of_fn
  25:     0x7f30315fde2e - rustc_hir_analysis[4501648514be1474]::collect::infer_return_ty_for_fn_sig
  26:     0x7f30315fd0bc - rustc_hir_analysis[4501648514be1474]::collect::fn_sig
  27:     0x7f30315f6585 - rustc_query_impl[c94a6fda47c12dc7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c94a6fda47c12dc7]::query_impl::fn_sig::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8ef003f1e476f553]::query::erase::Erased<[u8; 24usize]>>
  28:     0x7f30315f6555 - <rustc_query_impl[c94a6fda47c12dc7]::query_impl::fn_sig::dynamic_query::{closure#2} as core[d0d16022f048c5c1]::ops::function::FnOnce<(rustc_middle[8ef003f1e476f553]::ty::context::TyCtxt, rustc_span[fb1dd2d991832768]::def_id::DefId)>>::call_once
  29:     0x7f30315f5b93 - rustc_query_system[aa09d2e731c481a9]::query::plumbing::try_execute_query::<rustc_query_impl[c94a6fda47c12dc7]::DynamicConfig<rustc_query_system[aa09d2e731c481a9]::query::caches::DefaultCache<rustc_span[fb1dd2d991832768]::def_id::DefId, rustc_middle[8ef003f1e476f553]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[c94a6fda47c12dc7]::plumbing::QueryCtxt, false>
  30:     0x7f30315f55af - rustc_query_impl[c94a6fda47c12dc7]::query_impl::fn_sig::get_query_non_incr::__rust_end_short_backtrace
  31:     0x7f30318cca04 - <rustc_hir_analysis[4501648514be1474]::collect::CollectItemTypesVisitor as rustc_hir[200e79af71247e2]::intravisit::Visitor>::visit_item
  32:     0x7f30318caa3d - rustc_hir_analysis[4501648514be1474]::collect::collect_mod_item_types
  33:     0x7f30318ca9d1 - rustc_query_impl[c94a6fda47c12dc7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c94a6fda47c12dc7]::query_impl::collect_mod_item_types::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8ef003f1e476f553]::query::erase::Erased<[u8; 0usize]>>
  34:     0x7f3031db3b0f - rustc_query_system[aa09d2e731c481a9]::query::plumbing::try_execute_query::<rustc_query_impl[c94a6fda47c12dc7]::DynamicConfig<rustc_query_system[aa09d2e731c481a9]::query::caches::DefaultCache<rustc_span[fb1dd2d991832768]::def_id::LocalModDefId, rustc_middle[8ef003f1e476f553]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[c94a6fda47c12dc7]::plumbing::QueryCtxt, false>
  35:     0x7f3031db3643 - rustc_query_impl[c94a6fda47c12dc7]::query_impl::collect_mod_item_types::get_query_non_incr::__rust_end_short_backtrace
  36:     0x7f303160359f - rustc_hir_analysis[4501648514be1474]::check_crate
  37:     0x7f3032035ed9 - rustc_interface[82ea3601d258e7e1]::passes::analysis
  38:     0x7f3032035b1f - rustc_query_impl[c94a6fda47c12dc7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c94a6fda47c12dc7]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8ef003f1e476f553]::query::erase::Erased<[u8; 1usize]>>
  39:     0x7f30323a55a4 - rustc_query_system[aa09d2e731c481a9]::query::plumbing::try_execute_query::<rustc_query_impl[c94a6fda47c12dc7]::DynamicConfig<rustc_query_system[aa09d2e731c481a9]::query::caches::SingleCache<rustc_middle[8ef003f1e476f553]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[c94a6fda47c12dc7]::plumbing::QueryCtxt, false>
  40:     0x7f30323a5395 - rustc_query_impl[c94a6fda47c12dc7]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  41:     0x7f3032066f35 - rustc_interface[82ea3601d258e7e1]::interface::run_compiler::<core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>, rustc_driver_impl[d85864648598caad]::run_compiler::{closure#0}>::{closure#0}
  42:     0x7f30323db78a - std[735fb5c0265dddc8]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[82ea3601d258e7e1]::util::run_in_thread_with_globals<rustc_interface[82ea3601d258e7e1]::util::run_in_thread_pool_with_globals<rustc_interface[82ea3601d258e7e1]::interface::run_compiler<core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>, rustc_driver_impl[d85864648598caad]::run_compiler::{closure#0}>::{closure#0}, core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>>::{closure#0}, core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>>
  43:     0x7f30323db5b9 - <<std[735fb5c0265dddc8]::thread::Builder>::spawn_unchecked_<rustc_interface[82ea3601d258e7e1]::util::run_in_thread_with_globals<rustc_interface[82ea3601d258e7e1]::util::run_in_thread_pool_with_globals<rustc_interface[82ea3601d258e7e1]::interface::run_compiler<core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>, rustc_driver_impl[d85864648598caad]::run_compiler::{closure#0}>::{closure#0}, core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>>::{closure#0}, core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d0d16022f048c5c1]::result::Result<(), rustc_span[fb1dd2d991832768]::ErrorGuaranteed>>::{closure#1} as core[d0d16022f048c5c1]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  44:     0x7f302d3989f5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h29d894b5e99070ca
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/alloc/src/boxed.rs:2016:9
  45:     0x7f302d3989f5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hee2cef6a97c8434b
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/alloc/src/boxed.rs:2016:9
  46:     0x7f302d3989f5 - std::sys::unix::thread::Thread::new::thread_start::h2bd633f8986b3333
                               at /rustc/11035f9f52c65739d9613f5b9ecd332e9e640e71/library/std/src/sys/unix/thread.rs:108:17
  47:     0x7f302d1849eb - <unknown>
  48:     0x7f302d2087cc - <unknown>
  49:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.77.0-nightly (11035f9f5 2024-01-05) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [fn_sig] computing function signature of `call_this`
#1 [collect_mod_item_types] collecting item types in top-level module
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0405, E0601.
For more information about an error, try `rustc --explain E0405`.

```

</p>
</details>

<!--
query stack:
#0 [fn_sig] computing function signature of `call_this`
#1 [collect_mod_item_types] collecting item types in top-level module
#2 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_hir_analysis/src/astconv/lint.rs
+++ b/compiler/rustc_hir_analysis/src/astconv/lint.rs
@@ -122,7 +122,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                     ],
                     Applicability::MachineApplicable,
                 );
-            } else {
+            } else if diag.is_error() {
                 // We'll emit the object safety error already, with a structured suggestion.
                 diag.downgrade_to_delayed_bug();
             }
@@ -148,8 +148,10 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             }
             if !is_object_safe {
                 diag.note(format!("`{trait_name}` it is not object safe, so it can't be `dyn`"));
-                // We'll emit the object safety error already, with a structured suggestion.
-                diag.downgrade_to_delayed_bug();
+                if diag.is_error() {
+                    // We'll emit the object safety error already, with a structured suggestion.
+                    diag.downgrade_to_delayed_bug();
+                }
             } else {
                 let sugg = if let hir::TyKind::TraitObject([_, _, ..], _, _) = self_ty.kind {
                     // There are more than one trait bound, we need surrounding parentheses.

--- a/compiler/rustc_hir_analysis/src/astconv/lint.rs
+++ b/compiler/rustc_hir_analysis/src/astconv/lint.rs
@@ -78,14 +78,17 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
     fn maybe_lint_impl_trait(&self, self_ty: &hir::Ty<'_>, diag: &mut Diagnostic) -> bool {
         let tcx = self.tcx();
         let parent_id = tcx.hir().get_parent_item(self_ty.hir_id).def_id;
-        let (hir::Node::Item(hir::Item { kind: hir::ItemKind::Fn(sig, generics, _), .. })
-        | hir::Node::TraitItem(hir::TraitItem {
-            kind: hir::TraitItemKind::Fn(sig, _),
-            generics,
-            ..
-        })) = tcx.hir_node_by_def_id(parent_id)
-        else {
-            return false;
+        let (sig, generics, owner) = match tcx.hir_node_by_def_id(parent_id) {
+            hir::Node::Item(hir::Item { kind: hir::ItemKind::Fn(sig, generics, _), .. }) => {
+                (sig, generics, None)
+            }
+            hir::Node::TraitItem(hir::TraitItem {
+                kind: hir::TraitItemKind::Fn(sig, _),
+                generics,
+                owner_id,
+                ..
+            }) => (sig, generics, Some(tcx.parent(owner_id.to_def_id()))),
+            _ => return false,
         };
         let Ok(trait_name) = tcx.sess.source_map().span_to_snippet(self_ty.span) else {
             return false;
@@ -94,6 +97,11 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         let is_object_safe = match self_ty.kind {
             hir::TyKind::TraitObject(objects, ..) => {
                 objects.iter().all(|o| match o.trait_ref.path.res {
+                    Res::Def(DefKind::Trait, id) if Some(id) == owner => {
+                        // When we're dealing with a recursive trait, we don't want to downgrade
+                        // the error, so we consider them to be object safe always. (#119652)
+                        true
+                    }
                     Res::Def(DefKind::Trait, id) => tcx.check_is_object_safe(id),
                     _ => false,
                 })

--- a/tests/ui/object-safety/avoid-ice-on-warning-2.rs
+++ b/tests/ui/object-safety/avoid-ice-on-warning-2.rs
@@ -1,0 +1,9 @@
+fn id<F>(f: Copy) -> usize {
+//~^ WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+//~| WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+//~| ERROR the trait `Copy` cannot be made into an object
+    f()
+}
+fn main() {}

--- a/tests/ui/object-safety/avoid-ice-on-warning-2.stderr
+++ b/tests/ui/object-safety/avoid-ice-on-warning-2.stderr
@@ -1,0 +1,50 @@
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning-2.rs:1:13
+   |
+LL | fn id<F>(f: Copy) -> usize {
+   |             ^^^^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+   = note: `Copy` it is not object safe, so it can't be `dyn`
+   = note: `#[warn(bare_trait_objects)]` on by default
+help: use a new generic type parameter, constrained by `Copy`
+   |
+LL | fn id<F, T: Copy>(f: T) -> usize {
+   |        +++++++++     ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL | fn id<F>(f: impl Copy) -> usize {
+   |             ++++
+
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning-2.rs:1:13
+   |
+LL | fn id<F>(f: Copy) -> usize {
+   |             ^^^^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+   = note: `Copy` it is not object safe, so it can't be `dyn`
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+help: use a new generic type parameter, constrained by `Copy`
+   |
+LL | fn id<F, T: Copy>(f: T) -> usize {
+   |        +++++++++     ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL | fn id<F>(f: impl Copy) -> usize {
+   |             ++++
+
+error[E0038]: the trait `Copy` cannot be made into an object
+  --> $DIR/avoid-ice-on-warning-2.rs:1:13
+   |
+LL | fn id<F>(f: Copy) -> usize {
+   |             ^^^^ `Copy` cannot be made into an object
+   |
+   = note: the trait cannot be made into an object because it requires `Self: Sized`
+   = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+
+error: aborting due to 1 previous error; 2 warnings emitted
+
+For more information about this error, try `rustc --explain E0038`.

--- a/tests/ui/object-safety/avoid-ice-on-warning-3.rs
+++ b/tests/ui/object-safety/avoid-ice-on-warning-3.rs
@@ -1,0 +1,17 @@
+trait B { fn f(a: A) -> A; }
+//~^ WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN this is accepted in the current edition
+//~| WARN this is accepted in the current edition
+//~| WARN this is accepted in the current edition
+//~| ERROR the trait `A` cannot be made into an object
+trait A { fn g(b: B) -> B; }
+//~^ WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN this is accepted in the current edition
+//~| WARN this is accepted in the current edition
+//~| WARN this is accepted in the current edition
+//~| ERROR the trait `B` cannot be made into an object
+fn main() {}

--- a/tests/ui/object-safety/avoid-ice-on-warning-3.stderr
+++ b/tests/ui/object-safety/avoid-ice-on-warning-3.stderr
@@ -1,0 +1,148 @@
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning-3.rs:9:19
+   |
+LL | trait A { fn g(b: B) -> B; }
+   |                   ^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+   = note: `B` it is not object safe, so it can't be `dyn`
+   = note: `#[warn(bare_trait_objects)]` on by default
+help: use a new generic type parameter, constrained by `B`
+   |
+LL | trait A { fn g<T: B>(b: T) -> B; }
+   |               ++++++    ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL | trait A { fn g(b: impl B) -> B; }
+   |                   ++++
+
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning-3.rs:9:25
+   |
+LL | trait A { fn g(b: B) -> B; }
+   |                         ^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+help: `B` is not object safe, use `impl B` to return an opaque type, as long as you return a single underlying type
+   |
+LL | trait A { fn g(b: B) -> impl B; }
+   |                         ++++
+
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning-3.rs:1:19
+   |
+LL | trait B { fn f(a: A) -> A; }
+   |                   ^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+   = note: `A` it is not object safe, so it can't be `dyn`
+help: use a new generic type parameter, constrained by `A`
+   |
+LL | trait B { fn f<T: A>(a: T) -> A; }
+   |               ++++++    ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL | trait B { fn f(a: impl A) -> A; }
+   |                   ++++
+
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning-3.rs:1:25
+   |
+LL | trait B { fn f(a: A) -> A; }
+   |                         ^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+help: `A` is not object safe, use `impl A` to return an opaque type, as long as you return a single underlying type
+   |
+LL | trait B { fn f(a: A) -> impl A; }
+   |                         ++++
+
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning-3.rs:1:19
+   |
+LL | trait B { fn f(a: A) -> A; }
+   |                   ^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+   = note: `A` it is not object safe, so it can't be `dyn`
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+help: use a new generic type parameter, constrained by `A`
+   |
+LL | trait B { fn f<T: A>(a: T) -> A; }
+   |               ++++++    ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL | trait B { fn f(a: impl A) -> A; }
+   |                   ++++
+
+error[E0038]: the trait `A` cannot be made into an object
+  --> $DIR/avoid-ice-on-warning-3.rs:1:19
+   |
+LL | trait B { fn f(a: A) -> A; }
+   |                   ^ `A` cannot be made into an object
+   |
+note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+  --> $DIR/avoid-ice-on-warning-3.rs:9:14
+   |
+LL | trait A { fn g(b: B) -> B; }
+   |       -      ^ ...because associated function `g` has no `self` parameter
+   |       |
+   |       this trait cannot be made into an object...
+help: consider turning `g` into a method by giving it a `&self` argument
+   |
+LL | trait A { fn g(&self, b: B) -> B; }
+   |                ++++++
+help: alternatively, consider constraining `g` so it does not apply to trait objects
+   |
+LL | trait A { fn g(b: B) -> B where Self: Sized; }
+   |                           +++++++++++++++++
+
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning-3.rs:9:19
+   |
+LL | trait A { fn g(b: B) -> B; }
+   |                   ^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+   = note: `B` it is not object safe, so it can't be `dyn`
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+help: use a new generic type parameter, constrained by `B`
+   |
+LL | trait A { fn g<T: B>(b: T) -> B; }
+   |               ++++++    ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL | trait A { fn g(b: impl B) -> B; }
+   |                   ++++
+
+error[E0038]: the trait `B` cannot be made into an object
+  --> $DIR/avoid-ice-on-warning-3.rs:9:19
+   |
+LL | trait A { fn g(b: B) -> B; }
+   |                   ^ `B` cannot be made into an object
+   |
+note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+  --> $DIR/avoid-ice-on-warning-3.rs:1:14
+   |
+LL | trait B { fn f(a: A) -> A; }
+   |       -      ^ ...because associated function `f` has no `self` parameter
+   |       |
+   |       this trait cannot be made into an object...
+help: consider turning `f` into a method by giving it a `&self` argument
+   |
+LL | trait B { fn f(&self, a: A) -> A; }
+   |                ++++++
+help: alternatively, consider constraining `f` so it does not apply to trait objects
+   |
+LL | trait B { fn f(a: A) -> A where Self: Sized; }
+   |                           +++++++++++++++++
+
+error: aborting due to 2 previous errors; 6 warnings emitted
+
+For more information about this error, try `rustc --explain E0038`.

--- a/tests/ui/object-safety/avoid-ice-on-warning.rs
+++ b/tests/ui/object-safety/avoid-ice-on-warning.rs
@@ -1,0 +1,6 @@
+fn call_this<F>(f: F) : Fn(&str) + call_that {}
+//~^ ERROR return types are denoted using `->`
+//~| ERROR cannot find trait `call_that` in this scope
+//~| WARN trait objects without an explicit `dyn` are deprecated
+//~| WARN this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+fn main() {}

--- a/tests/ui/object-safety/avoid-ice-on-warning.stderr
+++ b/tests/ui/object-safety/avoid-ice-on-warning.stderr
@@ -1,0 +1,29 @@
+error: return types are denoted using `->`
+  --> $DIR/avoid-ice-on-warning.rs:1:23
+   |
+LL | fn call_this<F>(f: F) : Fn(&str) + call_that {}
+   |                       ^ help: use `->` instead
+
+error[E0405]: cannot find trait `call_that` in this scope
+  --> $DIR/avoid-ice-on-warning.rs:1:36
+   |
+LL | fn call_this<F>(f: F) : Fn(&str) + call_that {}
+   |                                    ^^^^^^^^^ not found in this scope
+
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/avoid-ice-on-warning.rs:1:25
+   |
+LL | fn call_this<F>(f: F) : Fn(&str) + call_that {}
+   |                         ^^^^^^^^^^^^^^^^^^^^
+   |
+   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
+   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
+   = note: `#[warn(bare_trait_objects)]` on by default
+help: `Fn(&str) + call_that` is not object safe, use `impl Fn(&str) + call_that` to return an opaque type, as long as you return a single underlying type
+   |
+LL | fn call_this<F>(f: F) : impl Fn(&str) + call_that {}
+   |                         ++++
+
+error: aborting due to 2 previous errors; 1 warning emitted
+
+For more information about this error, try `rustc --explain E0405`.

--- a/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021-without-dyn.rs
+++ b/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021-without-dyn.rs
@@ -1,0 +1,19 @@
+// edition:2021
+#![allow(bare_trait_objects)]
+trait A: Sized {
+    fn f(a: A) -> A;
+    //~^ ERROR trait objects must include the `dyn` keyword
+    //~| ERROR trait objects must include the `dyn` keyword
+}
+trait B {
+    fn f(a: B) -> B;
+    //~^ ERROR trait objects must include the `dyn` keyword
+    //~| ERROR trait objects must include the `dyn` keyword
+}
+trait C {
+    fn f(&self, a: C) -> C;
+    //~^ ERROR trait objects must include the `dyn` keyword
+    //~| ERROR trait objects must include the `dyn` keyword
+}
+
+fn main() {}

--- a/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021-without-dyn.rs
+++ b/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021-without-dyn.rs
@@ -4,11 +4,15 @@ trait A: Sized {
     fn f(a: A) -> A;
     //~^ ERROR trait objects must include the `dyn` keyword
     //~| ERROR trait objects must include the `dyn` keyword
+    //~| ERROR associated item referring to unboxed trait object for its own trait
+    //~| ERROR the trait `A` cannot be made into an object
 }
 trait B {
     fn f(a: B) -> B;
     //~^ ERROR trait objects must include the `dyn` keyword
     //~| ERROR trait objects must include the `dyn` keyword
+    //~| ERROR associated item referring to unboxed trait object for its own trait
+    //~| ERROR the trait `B` cannot be made into an object
 }
 trait C {
     fn f(&self, a: C) -> C;

--- a/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021-without-dyn.stderr
+++ b/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021-without-dyn.stderr
@@ -1,0 +1,105 @@
+error[E0782]: trait objects must include the `dyn` keyword
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:4:13
+   |
+LL |     fn f(a: A) -> A;
+   |             ^
+   |
+help: use a new generic type parameter, constrained by `A`
+   |
+LL |     fn f<T: A>(a: T) -> A;
+   |         ++++++    ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL |     fn f(a: impl A) -> A;
+   |             ++++
+help: alternatively, use a trait object to accept any type that implements `A`, accessing its methods at runtime using dynamic dispatch
+   |
+LL |     fn f(a: &dyn A) -> A;
+   |             ++++
+
+error[E0782]: trait objects must include the `dyn` keyword
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:4:19
+   |
+LL |     fn f(a: A) -> A;
+   |                   ^
+   |
+help: use `impl A` to return an opaque type, as long as you return a single underlying type
+   |
+LL |     fn f(a: A) -> impl A;
+   |                   ++++
+help: alternatively, you can return an owned trait object
+   |
+LL |     fn f(a: A) -> Box<dyn A>;
+   |                   +++++++  +
+
+error[E0782]: trait objects must include the `dyn` keyword
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:9:13
+   |
+LL |     fn f(a: B) -> B;
+   |             ^
+   |
+help: use a new generic type parameter, constrained by `B`
+   |
+LL |     fn f<T: B>(a: T) -> B;
+   |         ++++++    ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL |     fn f(a: impl B) -> B;
+   |             ++++
+help: alternatively, use a trait object to accept any type that implements `B`, accessing its methods at runtime using dynamic dispatch
+   |
+LL |     fn f(a: &dyn B) -> B;
+   |             ++++
+
+error[E0782]: trait objects must include the `dyn` keyword
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:9:19
+   |
+LL |     fn f(a: B) -> B;
+   |                   ^
+   |
+help: use `impl B` to return an opaque type, as long as you return a single underlying type
+   |
+LL |     fn f(a: B) -> impl B;
+   |                   ++++
+help: alternatively, you can return an owned trait object
+   |
+LL |     fn f(a: B) -> Box<dyn B>;
+   |                   +++++++  +
+
+error[E0782]: trait objects must include the `dyn` keyword
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:14:20
+   |
+LL |     fn f(&self, a: C) -> C;
+   |                    ^
+   |
+help: use a new generic type parameter, constrained by `C`
+   |
+LL |     fn f<T: C>(&self, a: T) -> C;
+   |         ++++++           ~
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL |     fn f(&self, a: impl C) -> C;
+   |                    ++++
+help: alternatively, use a trait object to accept any type that implements `C`, accessing its methods at runtime using dynamic dispatch
+   |
+LL |     fn f(&self, a: &dyn C) -> C;
+   |                    ++++
+
+error[E0782]: trait objects must include the `dyn` keyword
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:14:26
+   |
+LL |     fn f(&self, a: C) -> C;
+   |                          ^
+   |
+help: use `impl C` to return an opaque type, as long as you return a single underlying type
+   |
+LL |     fn f(&self, a: C) -> impl C;
+   |                          ++++
+help: alternatively, you can return an owned trait object
+   |
+LL |     fn f(&self, a: C) -> Box<dyn C>;
+   |                          +++++++  +
+
+error: aborting due to 6 previous errors
+
+For more information about this error, try `rustc --explain E0782`.

--- a/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021-without-dyn.stderr
+++ b/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021-without-dyn.stderr
@@ -1,3 +1,65 @@
+error: associated item referring to unboxed trait object for its own trait
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:4:13
+   |
+LL | trait A: Sized {
+   |       - in this trait
+LL |     fn f(a: A) -> A;
+   |             ^     ^
+   |
+help: you might have meant to use `Self` to refer to the implementing type
+   |
+LL |     fn f(a: Self) -> Self;
+   |             ~~~~     ~~~~
+
+error[E0038]: the trait `A` cannot be made into an object
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:4:13
+   |
+LL |     fn f(a: A) -> A;
+   |             ^ `A` cannot be made into an object
+   |
+note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:3:10
+   |
+LL | trait A: Sized {
+   |       -  ^^^^^ ...because it requires `Self: Sized`
+   |       |
+   |       this trait cannot be made into an object...
+
+error: associated item referring to unboxed trait object for its own trait
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:11:13
+   |
+LL | trait B {
+   |       - in this trait
+LL |     fn f(a: B) -> B;
+   |             ^     ^
+   |
+help: you might have meant to use `Self` to refer to the implementing type
+   |
+LL |     fn f(a: Self) -> Self;
+   |             ~~~~     ~~~~
+
+error[E0038]: the trait `B` cannot be made into an object
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:11:13
+   |
+LL |     fn f(a: B) -> B;
+   |             ^ `B` cannot be made into an object
+   |
+note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:11:8
+   |
+LL | trait B {
+   |       - this trait cannot be made into an object...
+LL |     fn f(a: B) -> B;
+   |        ^ ...because associated function `f` has no `self` parameter
+help: consider turning `f` into a method by giving it a `&self` argument
+   |
+LL |     fn f(&self, a: B) -> B;
+   |          ++++++
+help: alternatively, consider constraining `f` so it does not apply to trait objects
+   |
+LL |     fn f(a: B) -> B where Self: Sized;
+   |                     +++++++++++++++++
+
 error[E0782]: trait objects must include the `dyn` keyword
   --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:4:13
    |
@@ -33,7 +95,7 @@ LL |     fn f(a: A) -> Box<dyn A>;
    |                   +++++++  +
 
 error[E0782]: trait objects must include the `dyn` keyword
-  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:9:13
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:11:13
    |
 LL |     fn f(a: B) -> B;
    |             ^
@@ -52,7 +114,7 @@ LL |     fn f(a: &dyn B) -> B;
    |             ++++
 
 error[E0782]: trait objects must include the `dyn` keyword
-  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:9:19
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:11:19
    |
 LL |     fn f(a: B) -> B;
    |                   ^
@@ -67,7 +129,7 @@ LL |     fn f(a: B) -> Box<dyn B>;
    |                   +++++++  +
 
 error[E0782]: trait objects must include the `dyn` keyword
-  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:14:20
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:18:20
    |
 LL |     fn f(&self, a: C) -> C;
    |                    ^
@@ -86,7 +148,7 @@ LL |     fn f(&self, a: &dyn C) -> C;
    |                    ++++
 
 error[E0782]: trait objects must include the `dyn` keyword
-  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:14:26
+  --> $DIR/object-unsafe-trait-should-use-self-2021-without-dyn.rs:18:26
    |
 LL |     fn f(&self, a: C) -> C;
    |                          ^
@@ -100,6 +162,7 @@ help: alternatively, you can return an owned trait object
 LL |     fn f(&self, a: C) -> Box<dyn C>;
    |                          +++++++  +
 
-error: aborting due to 6 previous errors
+error: aborting due to 10 previous errors
 
-For more information about this error, try `rustc --explain E0782`.
+Some errors have detailed explanations: E0038, E0782.
+For more information about an error, try `rustc --explain E0038`.

--- a/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021.rs
+++ b/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021.rs
@@ -1,0 +1,17 @@
+// edition:2021
+#![allow(bare_trait_objects)]
+trait A: Sized {
+    fn f(a: dyn A) -> dyn A;
+    //~^ ERROR associated item referring to unboxed trait object for its own trait
+    //~| ERROR the trait `A` cannot be made into an object
+}
+trait B {
+    fn f(a: dyn B) -> dyn B;
+    //~^ ERROR associated item referring to unboxed trait object for its own trait
+    //~| ERROR the trait `B` cannot be made into an object
+}
+trait C {
+    fn f(&self, a: dyn C) -> dyn C;
+}
+
+fn main() {}

--- a/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021.stderr
+++ b/tests/ui/suggestions/object-unsafe-trait-should-use-self-2021.stderr
@@ -1,0 +1,65 @@
+error: associated item referring to unboxed trait object for its own trait
+  --> $DIR/object-unsafe-trait-should-use-self-2021.rs:4:13
+   |
+LL | trait A: Sized {
+   |       - in this trait
+LL |     fn f(a: dyn A) -> dyn A;
+   |             ^^^^^     ^^^^^
+   |
+help: you might have meant to use `Self` to refer to the implementing type
+   |
+LL |     fn f(a: Self) -> Self;
+   |             ~~~~     ~~~~
+
+error[E0038]: the trait `A` cannot be made into an object
+  --> $DIR/object-unsafe-trait-should-use-self-2021.rs:4:13
+   |
+LL |     fn f(a: dyn A) -> dyn A;
+   |             ^^^^^ `A` cannot be made into an object
+   |
+note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+  --> $DIR/object-unsafe-trait-should-use-self-2021.rs:3:10
+   |
+LL | trait A: Sized {
+   |       -  ^^^^^ ...because it requires `Self: Sized`
+   |       |
+   |       this trait cannot be made into an object...
+
+error: associated item referring to unboxed trait object for its own trait
+  --> $DIR/object-unsafe-trait-should-use-self-2021.rs:9:13
+   |
+LL | trait B {
+   |       - in this trait
+LL |     fn f(a: dyn B) -> dyn B;
+   |             ^^^^^     ^^^^^
+   |
+help: you might have meant to use `Self` to refer to the implementing type
+   |
+LL |     fn f(a: Self) -> Self;
+   |             ~~~~     ~~~~
+
+error[E0038]: the trait `B` cannot be made into an object
+  --> $DIR/object-unsafe-trait-should-use-self-2021.rs:9:13
+   |
+LL |     fn f(a: dyn B) -> dyn B;
+   |             ^^^^^ `B` cannot be made into an object
+   |
+note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+  --> $DIR/object-unsafe-trait-should-use-self-2021.rs:9:8
+   |
+LL | trait B {
+   |       - this trait cannot be made into an object...
+LL |     fn f(a: dyn B) -> dyn B;
+   |        ^ ...because associated function `f` has no `self` parameter
+help: consider turning `f` into a method by giving it a `&self` argument
+   |
+LL |     fn f(&self, a: dyn B) -> dyn B;
+   |          ++++++
+help: alternatively, consider constraining `f` so it does not apply to trait objects
+   |
+LL |     fn f(a: dyn B) -> dyn B where Self: Sized;
+   |                             +++++++++++++++++
+
+error: aborting due to 4 previous errors
+
+For more information about this error, try `rustc --explain E0038`.

